### ISSUE TITLE
How to observe velocity?

### ISSUE THREAD

**Issue by gzrjzcx** [Unity-Technologies/ml-agents#2387]
Hi there, 
I am training an agent to finish a stack task like the image:
<img width="479" alt="屏幕快照 2019-08-04 下午3 28 43" src="https://user-images.githubusercontent.com/33338567/62424869-b2271480-b6cc-11e9-8cd3-4a33b855bd57.png">
  
After several attempts, the model is able to stack the block to some extent. However, it is not accurate, although the cumulative reward reached 0.8(scaled to 1). It looks like that the model is not able to consider the velocity of the swinging sling and swaying column. Therefore, I want to add the velocity of the sling and the column as the observation.
  
The problem is that the swing of the sling and the sway of the column are controlled via varying the localposition and localrotation, not velocity. Then, how can I observe the velocity of both? 
  
Plus, I am observing the localposition and localrotation now instead of velocity. Can I calculate the speed manually and add them to observation? Or can I stack the observations by `stacked vectors` , then the agent is able to know the position and rotation of previous several frame, thus has the notion of the velocity?

**Comment by cgiles** [comment]
Velocity is equal to distance/time, so if you do position(t)-position(t-1) t being the instant, you will get your velocity at this instant. You should store your previous position, then subtract it to your actual position, and you will get your velocity.

**Comment by awjuliani** [comment]
Hi @gzrjzcx,

As @cgiles mentions, I would recommend calculating it directly using differences between positions, and then passing those variables as observations to the agent.

**Comment by gzrjzcx** [comment]
Hi @awjuliani,
I have calculated it directly now. But I want to know is this possible to observe the velocity via stacking vectors?  Because I have checked your `Tennis` environment, you have observed the velocity directly, but you still use the 3 stacked vectors.
  
In addition,  is the `--load` command used for continuing training? I find that using the `--load` for curriculum learning, although I have specify the lesson to the corresponding lesson at the end point, the mean reward and return(specify `--debug`) still decreased dramatically.
In my case, at the end of the first 5M steps, the return is 0.26 and the mean reward for the last 1000 steps is about 0.6. However, when I started the following 5M steps, the mean reward at the 5M + 1000 steps decreased to -0.1 and the returns decreased to the 0.06. 
I mean the problem is that it cannot restore to the parameters at the end  of the last training. Or why the mean reward and return decreased?

**Comment by gzrjzcx** [comment]
Hi @awjuliani,
In this case, the first training ended at 5M step. Then I continue the training by `--load` without any vary in the config file except the increasing of `max_steps` from 5M to 10M. 
Then you can see the mean reward decreased dramatically from the start of the following training(i.e. 5M+1 step).  Although then it increases finally, it is not same as a direct training of 10M. 
  
<img width="1066" alt="屏幕快照 2019-08-07 下午1 51 01" src="https://user-images.githubusercontent.com/33338567/62624715-949bba00-b91b-11e9-82e7-0154aa04452f.png">

And I have checked the statistics, the learning rate looks weird and I can't explain why it increased. Is it relevant to the decrease? 
  
<img width="1053" alt="屏幕快照 2019-08-07 下午1 51 16" src="https://user-images.githubusercontent.com/33338567/62624718-99f90480-b91b-11e9-8ff8-48a4d7e26414.png">

**Comment by awjuliani** [comment]
Hi @gzrjzcx 

The learning rate is set to a ratio of the total steps based on the completed steps. When you increase the total steps, you increase the learning rate again, hence the nature of the curve you see. Going along with this is a decrease in stability of learning, which you see in your performance results as well.

**Comment by gzrjzcx** [comment]
Hi @awjuliani,
So I can't continue training in this case right? Because it has reached the max_step, which means the training has finished. If I change the max_step, it cannot restore to the previous situation right?
  
BTW, I have calculated velocity directly now. But I want to know is this possible to observe the velocity via stacking vectors? Because I have checked your Tennis environment, you have observed the velocity directly, but you still use the 3 stacked vectors.
